File: backend/api/routers/polls.py
from fastapi import APIRouter, Depends, status, HTTPException
from typing import Literal, List
from sse_starlette.sse import EventSourceResponse
from ..dependencies import get_poll_service, get_vote_service, get_poll_event_bus
from ..services.poll_service import PollService
from ..services.vote_service import VoteService
from ..auth import get_current_user_simple, get_optional_user_simple
from ..schemas.poll import PollCreate, Poll
from ..schemas.user import User
from ..events import PollEventBus
from pydantic import BaseModel
import json


router = APIRouter(prefix="/polls", tags=["Polls"])


class PollListRequest(BaseModel):
    sort_by: Literal["created_at", "likes"] = "created_at"
    limit: int = 50
    offset: int = 0


@router.post("/", response_model=Poll, status_code=status.HTTP_201_CREATED)
async def create_poll(
    payload: PollCreate,
    current_user: User = Depends(get_current_user_simple),
    poll_service: PollService = Depends(get_poll_service),
):
    return await poll_service.create_poll(user_id=current_user.id, payload=payload)


@router.delete("/{poll_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_poll(
    poll_id: int,
    current_user: User = Depends(get_current_user_simple),
    poll_service: PollService = Depends(get_poll_service),
):
    poll = await poll_service.get_poll(poll_id)
    if (poll.created_by != current_user.id) and (not current_user.is_admin):
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="Not allowed to delete this poll")
    await poll_service.delete_poll(poll_id=poll_id)
    return None


@router.post("/list", response_model=List[Poll])
async def list_polls(
    body: PollListRequest,
    poll_service: PollService = Depends(get_poll_service),
    vote_service: VoteService = Depends(get_vote_service),
    maybe_user: User | None = Depends(get_optional_user_simple),
):
    polls = await poll_service.list_polls_ranked(by=body.sort_by, limit=body.limit, offset=body.offset)
    if not maybe_user:
        return polls
    poll_ids = [p.id for p in polls]
    mapping = await vote_service.list_user_votes_for_polls(user_id=maybe_user.id, poll_ids=poll_ids)
    # Ensure my_vote_option_id is included in response
    enriched: List[Poll] = []
    for p in polls:
        enriched.append(Poll.from_orm(p).copy(update={"my_vote_option_id": mapping.get(p.id)}))
    return enriched


@router.post("/mine", response_model=List[Poll])
async def list_my_polls(
    body: PollListRequest,
    current_user: User = Depends(get_current_user_simple),
    poll_service: PollService = Depends(get_poll_service),
    vote_service: VoteService = Depends(get_vote_service),
):
    polls = await poll_service.list_polls_by_user(user_id=current_user.id, by=body.sort_by, limit=body.limit, offset=body.offset)
    poll_ids = [p.id for p in polls]
    mapping = await vote_service.list_user_votes_for_polls(user_id=current_user.id, poll_ids=poll_ids)
    enriched: List[Poll] = []
    for p in polls:
        enriched.append(Poll.from_orm(p).copy(update={"my_vote_option_id": mapping.get(p.id)}))
    return enriched


@router.post("/{poll_id}/like", response_model=Poll)
async def like_poll(
    poll_id: int,
    current_user: User = Depends(get_current_user_simple),
    poll_service: PollService = Depends(get_poll_service),
):
    return await poll_service.like_poll(poll_id=poll_id, user_id=current_user.id)


@router.post("/{poll_id}/dislike", response_model=Poll)
async def dislike_poll(
    poll_id: int,
    current_user: User = Depends(get_current_user_simple),
    poll_service: PollService = Depends(get_poll_service),
):
    return await poll_service.dislike_poll(poll_id=poll_id, user_id=current_user.id)


@router.get("/stream")
async def stream_poll_updates(
    event_bus: PollEventBus = Depends(get_poll_event_bus),
):
    """
    Server-Sent Events endpoint for real-time poll updates.
    Broadcasts poll_created, poll_updated, and poll_deleted events.
    """
    async def event_generator():
        async for event in event_bus.subscribe():
            # Format: event_type and data fields for SSE
            yield {
                "event": event.event_type,
                "data": json.dumps(event.payload),
            }
    
    return EventSourceResponse(event_generator())

File: frontend/lib/polls.ts
import { apiFetch } from "./api-client";
import type { AuthUser } from "./auth";

export type PollSortBy = "created_at" | "likes";

export interface PollOption {
  id: number;
  text: string;
  votes: number;
}

export interface PollOptionPayload {
  text: string;
}

export interface Poll {
  id: number;
  title: string;
  description: string | null;
  likes: number;
  dislikes: number;
  poll_expires_at: string | null;
  created_at: string;
  updated_at: string | null;
  created_by: number;
  creator: AuthUser;
  options: PollOption[];
  liked_by: AuthUser[];
  disliked_by: AuthUser[];
  my_vote_option_id?: number | null;
}

export interface CreatePollPayload {
  title: string;
  description?: string | null;
  poll_expires_at?: string | null;
  options: PollOptionPayload[];
}

export interface PollListRequest {
  sort_by: PollSortBy;
  limit?: number;
  offset?: number;
}

export interface VotePayload {
  poll_id: number;
  option_id: number;
}

export interface Vote {
  id: number;
  user_id: number;
  poll_id: number;
  option_id: number;
  created_at: string;
}

export async function createPoll(payload: CreatePollPayload, token: string) {
  return apiFetch<Poll>("/api/polls/", {
    method: "POST",
    body: payload,
    token,
  });
}

export async function listPolls(payload: PollListRequest, token?: string | null) {
  return apiFetch<Poll[]>("/api/polls/list", {
    method: "POST",
    body: payload,
    token: token ?? undefined,
  });
}

export async function listMyPolls(payload: PollListRequest, token: string) {
  return apiFetch<Poll[]>("/api/polls/mine", {
    method: "POST",
    body: payload,
    token,
  });
}

export async function deletePoll(pollId: number | string, token: string) {
  await apiFetch<void>(`/api/polls/${pollId}`, {
    method: "DELETE",
    token,
  });
}

export async function likePoll(pollId: number | string, token: string) {
  return apiFetch<Poll>(`/api/polls/${pollId}/like`, {
    method: "POST",
    token,
  });
}

export async function dislikePoll(pollId: number | string, token: string) {
  return apiFetch<Poll>(`/api/polls/${pollId}/dislike`, {
    method: "POST",
    token,
  });
}

export async function castVote(payload: VotePayload, token: string) {
  return apiFetch<Poll>("/api/votes/", {
    method: "POST",
    body: payload,
    token,
  });
}

// Server-Sent Events (SSE) for real-time poll updates
export interface PollStreamEvent {
  type: "poll_created" | "poll_updated" | "poll_deleted";
  poll?: Poll;
  poll_id?: number;
}

export function createPollStream(
  onEvent: (event: PollStreamEvent) => void,
  onError?: (error: Error) => void
): EventSource {
  const eventSource = new EventSource(`${process.env.NEXT_PUBLIC_API_URL || "http://localhost:8000"}/api/polls/stream`);

  eventSource.addEventListener("poll_created", (event) => {
    try {
      const data = JSON.parse(event.data);
      onEvent({ type: "poll_created", poll: data.poll });
    } catch (error) {
      onError?.(error as Error);
    }
  });

  eventSource.addEventListener("poll_updated", (event) => {
    try {
      const data = JSON.parse(event.data);
      onEvent({ type: "poll_updated", poll: data.poll });
    } catch (error) {
      onError?.(error as Error);
    }
  });

  eventSource.addEventListener("poll_deleted", (event) => {
    try {
      const data = JSON.parse(event.data);
      onEvent({ type: "poll_deleted", poll_id: data.poll_id });
    } catch (error) {
      onError?.(error as Error);
    }
  });

  eventSource.onerror = (event) => {
    onError?.(new Error("EventSource connection error"));
  };

  return eventSource;
}

